src/backend/catalog-builder/npm-downloads.lambda-shared.ts

```diff
@@ -0,0 +1,150 @@
+import type { Got } from 'got/dist/source/types';
+
+export enum NpmDownloadsPeriod {
+  /**
+   * Gets downloads for the last available day. In practice, this will usually
+   * be "yesterday" (in GMT) but if stats for that day have not yet landed, it
+   * will be the day before.
+   */
+  LAST_DAY = 'last-day',
+
+  /**
+   * Gets downloads for the last 7 available days.
+   */
+  LAST_WEEK = 'last-week',
+
+  /**
+   * Gets downloads for the last 30 available days.
+   */
+  LAST_MONTH = 'last-month'
+}
+
+export interface NpmDownloadsEntry {
+  readonly downloads: number;
+  readonly start: string;
+  readonly end: string;
+  readonly package: string;
+}
+
+export type NpmDownloadsOutput = { [key: string]: NpmDownloadsEntry };
+
+export interface NpmDownloadsOptions {
+  /**
+   * The period to query for package download counts.
+   * @default NpmDownloadsPeriod.LAST_WEEK
+   */
+  readonly period?: NpmDownloadsPeriod;
+
+  /**
+   * Throw an error when any package's download metrics are not available.
+   * @default true
+   */
+  readonly throwErrors?: boolean;
+}
+
+export class NpmDownloadsClient {
+  public static readonly NPM_DOWNLOADS_API_URL = 'https://api.npmjs.org/downloads/point';
+  public static readonly MAX_PACKAGES_PER_QUERY = 128; // this is a limitation from npm's API! do not change
+
+  private readonly got: Got;
+
+  constructor(gotService: Got) {
+    this.got = gotService;
+  }
+
+  private async getDownloadsRaw(
+    packages: string[],
+    period: NpmDownloadsPeriod,
+    throwErrors: boolean,
+  ): Promise<NpmDownloadsOutput> {
+    if (packages.length > NpmDownloadsClient.MAX_PACKAGES_PER_QUERY) {
+      throw new Error(`Too many packages were provided (max: ${NpmDownloadsClient.MAX_PACKAGES_PER_QUERY})`);
+    }
+    if (packages.length === 0) return {};
+
+    console.log(`Querying NPM for packages: [${packages.join(',')}]...`);
+    const result = await this.got(`${NpmDownloadsClient.NPM_DOWNLOADS_API_URL}/${period}/${packages.join(',')}`, {
+      timeout: 5 * 1000, // 30 seconds
+    }).catch((err) => {
+      if (throwErrors) {
+        throw err;
+      } else {
+        return { body: JSON.stringify({ error: JSON.stringify(err) }) };
+      }
+    });
+
+    const data = JSON.parse(result.body);
+
+    // single package query error
+    if ('error' in data) {
+      if (throwErrors) {
+        throw new Error(`Could not retrieve download metrics: ${data.error}`);
+      } else {
+        console.error(`Could not retrieve download metrics: ${data.error}`);
+        return {};
+      }
+    }
+    // bulk package query error
+    for (const key of Object.keys(data)) {
+      if (!data[key]) {
```

Comment:
I think we should go for an existence check here... If the value is `number`, since `0` is falsey in JS... that'd get us some surprises...

```suggestion
      if (!(key in data)) {
```

Reply:
Updated.